### PR TITLE
Subsurface scattering support in Blinn-Phong material 

### PR DESCRIPTION
**Pull request by daoshengmu**
Per #9249, according to the approach from [GDC 2011 – Approximating Translucency for a Fast, Cheap and Convincing Subsurface Scattering Look](url), implementing the subsurface scattering shader for PBR material.

### CODE REVIEW

**Comment by daoshengmu** [comment]
screen shot 2018-03-07 at 1 47 31 pm
The demo.

**Comment by WestLangley** [comment]
Thanks for this! I have a few issues, though...

1. I see this is not a PBR technique,  nor is it based on any PBR models AFAIK, and therefore should not be an extension of `MeshStandardMaterial` -- maybe `MeshPhongMaterial`,  otherwise just extend `ShaderMaterial`.

2. Also, this material requires a `thicknessTexture` most models do not have. Consequently, I do not think it should be a built-in material -- and certainly not until the renderer is modified so that it is material-agnostic. An alternate approach would be to use `NodeMaterial`, perhaps. In an event, it belongs as an example for now.

I feel fairly strongly about both of these points, but I am glad you did this, and would like to see your example included.

**Comment by daoshengmu** [comment]
@WestLangley Thanks for your comment. I was considering it should not extend `MeshStandardMaterial ` or `MeshPhongMaterial `, because it is just an additional diffuse light. It should not rely on PBR or Blinn-Phong lighting model. So, probably, `ShaderMaterial` is more wise to choose.

As the `ShaderMaterial` example as below,
```
var shader = THREE.ShaderLib[ 'physical' ];
var material = new THREE.ShaderMaterial( { fragmentShader: shader.fragmentShader, vertexShader: shader.vertexShader, uniforms: uniforms: THREE.UniformsUtils.merge( [
                                                THREE.ShaderLib[ 'physical' ].uniforms,
						THREE.UniformsLib[ "translucent" ]
					], lights: true } );

material.uniforms['thicknessMap'] = thicknessTexture;
...
```

In `lights_physical_pars_fragment.glsl`, `lights_phong_pars_fragment.glsl`, and `meshlambert_frag.glsl`, we can add
```
#ifdef USE_THICKNESSMAP
  RE_Direct_Scattering();
#endif
```
for diffuse scattering.

If it is workable for this approach, do we still need `NodeMaterial` or both of these two ways are needed?

**Comment by WestLangley** [comment]
>It should not rely on PBR or Blinn-Phong lighting model. 

Correct. It should not. It should not rely on the Physical shader, either.

>do we still need NodeMaterial

No, you do not have to use `NodeMaterial` if you do not want to. Consider following the example `webgl_materials_skin.html`. I do not think it should be a built-in material, so you do not need to modify the renderer.

**Comment by daoshengmu** [comment]
@WestLangley I have followed your suggestion to move the subsurface scattering shader to be `ShaderTranslucent` under the `examples/` folder, and giving demos for `Standard`, `Phong`, and `Lambert` lighting models.

Please give me some feedback to confirm if it can be merged. Thanks!

**Comment by daoshengmu** [comment]
screen shot 2018-03-13 at 12 13 53 am
The demos.

**Comment by WestLangley** [comment]
@daoshengmu Thanks for this!

As I said, since this technique is not physically based, I do not think you should be using the `MeshStandardMaterial` shader chunks. `MeshStandardMaterial` supports metals; metals are not translucent.

1. I suggest you have only one translucent material, and use the Phong shader chunks. Rename it `THREE.TranslucentShader`. You then only need one mesh in your example.

2. Remove the environment map from the example since it is not used in your material.

With those two simplifications, this should be ready to merge.

Your results should look similar the excellent example here: https://github.com/mrdoob/three.js/issues/9249.

**Comment by daoshengmu** [comment]
After do some adjustment of the material parameters, it looks much better.
screen shot 2018-03-13 at 11 10 46 am

@WestLangley IIUC, the difference between PBR and Blinn-Phong is the part of specular lighting. Despite the PBR defines metalness value in its formula, we still can give it zero to make it be non-metalness. PBR's specular lighting is more closer to the cases in the real world. Therefore, some materials like skin, wax still need it although they are not metal as well.

Besides, I am curious why you mentioned `environment map`? I don't see I am using `environment map` in my shader. Probably, I should remove `USE_SHADOWMAP` because I really don't want to do shadow map in my demo.

**Comment by WestLangley** [comment]
Your shader is not PBR, and should not be presented as such. I stand by my recommendations in my previous comment.

>Besides, I am curious why you mentioned environment map?

I meant remove `reflectionCube` and the scene background because the environment is neither reflected nor refracted in your shader.

**Comment by daoshengmu** [comment]
@WestLangley, Interesting. Maybe I misunderstand the PBR in three.js. If I would like to use PBR, which shader I should refer? the ‘MeshPhysicalMaterial’?

**Comment by WestLangley** [comment]
PBR is based on the physics of light. But the technique you are implementing is not based on the physics of light (subsurface scattering, in this case) -- it is a "Cheap and Convincing Subsurface Scattering Look".

If you want to mix PBR and this technique in your own projects, that is OK. But here, I do not want to give the impression that a shader is PBR when it is not.

Both `MeshStandardMaterial` and `MeshPhysicalMaterial` are PBR-based. If you would like to use built-in "shader chunks", I think it would be reasonable to use the shader chunks from `Phong`.

The light should be behind the object (backlit) to demonstrate the effect you are trying to incorporate into your shader.

**Comment by daoshengmu** [comment]
> If you want to mix PBR and this technique in your own projects, that is OK. But here, I do not want to give the impression that a shader is PBR when it is not.

It makes sense. I have followed your comments to fix my commits. Thanks for the suggestion.

> The light should be behind the object (backlit) to demonstrate the effect you are trying to incorporate into your shader.

I have one directional light and two point lights in my demo. The 1st point light (white color) who has a small radius and put in front of the model is for adding brightness to this model, it will not have scattering on the model's back side because of the point light's small radius. The 2nd point light (yellow color) I put behind the model with larger radius will generate scattering and be presented on the model's front side.

**Comment by Mugen87** [comment]
I don't think the following location is correct for the FBX file: `examples/obj/bunny/stanford-bunny.fbx`

Please use `examples/models/fbx/stanford-bunny.fbx`. Or instead of adding a new model, use one of the existing ones...

**Comment by daoshengmu** [comment]
@Mugen87 Thanks for the correction. I have moved the fbx model to `examples/models/fbx/stanford-bunny.fbx`

File: examples/webgl_materials_translucency.html [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+			uniforms[ "thicknessScale" ].value = 10.0;
+
+			var parameters = { fragmentShader: shader.fragmentShader, vertexShader: shader.vertexShader, uniforms: uniforms, lights: true };
+			var material = new THREE.ShaderMaterial( parameters );
+			material.extensions.derivatives = true;
+
+			// LOADER
+
+			var loader = new THREE.FBXLoader();
+			loader.load("models/fbx/stanford-bunny.fbx", function(geometry) {
```
Comment (WestLangley) [inline examples/webgl_materials_translucency.html]: `FBXLoader.load()` does not return a geometry.

File: examples/webgl_materials_translucency.html [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+				loadedModel(geometry, 1, material);
+			});
+		}
+
+		function loadedModel(mesh, scale, material) {
```
Comment (WestLangley) [inline examples/webgl_materials_translucency.html]: It is not a `mesh` either.

**Comment by WestLangley** [comment]
Thanks @daoshengmu. I think this can be merged, but the code is going to need cleanup, which can be done later.

/ping @Mugen87

**Comment by daoshengmu** [comment]
I have replaced the `geometry` and `mesh` with `object` according the code review comment.

**Comment by daoshengmu** [comment]
@Mugen87 It seems to have been a couple of days. I am curious when is the possible timeline to merge my patch?

Thanks!

**Review by Mugen87 — APPROVED** [review]

**Comment by daoshengmu** [comment]
@mrdoob I am interested if it is available to be merged. Thanks!

**Comment by WestLangley** [comment]
@daoshengmu This PR has an r92 milestone, which means if @mrdoob approves it, it will be merged before the next release. Thank you for being patient.

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**